data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/EnumDeserializer.java

```diff
@@ -0,0 +1,57 @@
+package org.opensearch.dataprepper.pipeline.parser;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.databind.BeanProperty;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.deser.ContextualDeserializer;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+
+/**
+ * This deserializer is used for any Enum classes when converting the pipeline configuration file into the plugin model classes
+ * @since 2.11
+ */
+public class EnumDeserializer extends JsonDeserializer<Enum<?>> implements ContextualDeserializer {
+
+    static final String INVALID_ENUM_VALUE_ERROR_FORMAT = "Invalid value \"%s\". Valid options include %s.";
+
+    private Class<?> enumClass;
+
+    public EnumDeserializer() {}
+
+    public EnumDeserializer(final Class<?> enumClass) {
+        if (!enumClass.isEnum()) {
+            throw new IllegalArgumentException("The provided class is not an enum: " + enumClass.getName());
+        }
+
+        this.enumClass = enumClass;
+    }
+    @Override
+    public Enum<?> deserialize(final JsonParser p, final DeserializationContext ctxt) throws IOException {
+        final JsonNode node = p.getCodec().readTree(p);
+        final String enumValue = node.asText();
+
+        for (Object enumConstant : enumClass.getEnumConstants()) {
+            if (enumConstant.toString().equalsIgnoreCase(enumValue)) {
```

Comment:
I don't think this is the right approach. We should be sticking with a single case to avoid multiple possible values for the same thing.

Rather, use the core Jackson deserializer to get the value. Or, use `@JsonCreator` which is what it uses.

Reply:
By the way, I think that is fine to make something like `enumConstant.toString().toLowerCase().equals(enumValue)` as a default behavior. But, we should still support `@JsonValue` and `@JsonCreator`. So, you could have logic like the following pseudo-code:

```
if(findAnnotedMethod(enumClass, JsonCreator.class)) {
  return useJsonCreator(enumClass);
} else {
  if( enumConstant.toString().toLowerCase().equals(enumValue) {
    return enumConstant;
  }
}
```

Reply:
Is there a problem with allowing both lowercase and uppercase? I think it is fair to assume that if the user inputs either, this is the intended choice. I have felt like it is not clear which value to put into the YAML before.

Reply:
We should avoid it because it ends up causing exactly this confusion. So rather than be so loose that anything goes, let's just require the correct input. We can consistently take in `lowercase-values` which is what our documentation tells people to do.

Also, this is so loose that we will allow `aLL-soRTs-of-VALues`.

Reply:
Can JsonValue just be used for this too? Not sure how we would use JsonCreator for this

Reply:
Right now the enums all use `@JsonCreator` to determine how to deserialize from JSON into POJO. The `@JsonValue` determines how to reverse this for the schemas. Maybe `@JsonValue` can do both, but I'm not sure.

This is the pattern:

https://github.com/opensearch-project/data-prepper/blob/aff479b61011b476a1ff87a2dda2ae73e83c8542/data-prepper-plugins/mutate-event-processors/src/main/java/org/opensearch/dataprepper/plugins/processor/mutateevent/TargetType.java#L52-L61